File: packages/sdk/src/agent.ts
/**
 * `ArpAgent` — the developer-facing entry point for building an ARP agent in
 * TypeScript. Wraps:
 *
 *   @kybernesis/arp-runtime     HTTP + DIDComm dispatch pipeline
 *   @kybernesis/arp-pdp         Cedar PDP
 *   @kybernesis/arp-transport   DIDComm envelopes
 *   @kybernesis/arp-registry    SQLite connection + revocation store
 *   @kybernesis/arp-audit       per-connection JSONL audit log
 *
 * Consumers only see the five integration points defined in
 * `ARP-installation-and-hosting.md §8`:
 *
 *   agent.check({ action, resource, context, connectionId })
 *   agent.egress({ data, connectionId, obligations })
 *   agent.onIncoming((task, ctx) => ...)
 *   agent.audit({ connectionId, decision, ... })
 *   agent.on('revocation' | 'rotation' | 'pairing', handler)
 *
 * Every framework adapter under `adapters/*` is expected to be a ~200 line
 * glue module that translates framework idioms (plugins, middleware,
 * decorators, graph nodes) into those five calls.
 */

import { createRequire } from 'node:module';
import { join } from 'node:path';
import {
  createRuntime,
  type DispatchHandler,
  type DispatchInput,
  type Runtime,
} from '@kybernesis/arp-runtime';
import { createInMemoryKeyStore } from '@kybernesis/arp-transport';
import { createResolver } from '@kybernesis/arp-resolver';
import type { Resolver } from '@kybernesis/arp-resolver';
import type { ConnectionRecord, Registry } from '@kybernesis/arp-registry';
import type { ConnectionToken, HandoffBundle, Obligation } from '@kybernesis/arp-spec';
import type { PdpDecision } from '@kybernesis/arp-pdp';
import { bootstrapSdk } from './bootstrap.js';
import { applyObligations } from './obligations.js';
import type {
  ArpAgentEvent,
  ArpAgentEventPayload,
  ArpAgentOptions,
  AuditEventInput,
  CheckInput,
  ConnectionAPI,
  EgressInput,
  InboundHandler,
  PairingEvent,
  PdpAPI,
  RegistryReadAPI,
  Resource,
  RevocationEvent,
  RotationEvent,
} from './types.js';

const requireFromHere = createRequire(import.meta.url);

type EventHandlers = {
  revocation: Array<(e: RevocationEvent) => void>;
  rotation: Array<(e: RotationEvent) => void>;
  pairing: Array<(e: PairingEvent) => void>;
};

export class ArpAgent {
  /** Agent DID. */
  public readonly did: string;
  /** Principal DID. */
  public readonly principalDid: string;
  /** Agent's public key (multibase). */
  public readonly publicKeyMultibase: string;

  /** Sub-APIs. */
  public readonly connections: ConnectionAPI;
  public readonly registry: RegistryReadAPI;
  public readonly pdp: PdpAPI;

  private readonly runtime: Runtime;
  private readonly now: () => number;
  private readonly events: EventHandlers;

  private constructor(params: {
    did: string;
    principalDid: string;
    publicKeyMultibase: string;
    runtime: Runtime;
    now: () => number;
  }) {
    this.did = params.did;
    this.principalDid = params.principalDid;
    this.publicKeyMultibase = params.publicKeyMultibase;
    this.runtime = params.runtime;
    this.now = params.now;
    this.events = { revocation: [], rotation: [], pairing: [] };

    const registry = this.runtime.registry;
    this.connections = buildConnectionAPI(registry, this.runtime, this.events, this.now);
    this.registry = buildRegistryReadAPI(registry);
    this.pdp = buildPdpAPI(this.runtime);
  }

  /**
   * Build an agent from a handoff bundle + options. Does not start the HTTP
   * server yet — call `agent.start()` once you've attached your handlers.
   */
  static async fromHandoff(
    handoffInput: HandoffBundle | string | Record<string, unknown>,
    options: ArpAgentOptions = {},
  ): Promise<ArpAgent> {
    const dataDir = options.dataDir ?? join(process.cwd(), '.arp-data');
    const now = options.now ?? (() => Date.now());

    const bootstrap = await bootstrapSdk({
      handoff: handoffInput,
      dataDir,
      ...(options.privateKey ? { privateKey: options.privateKey } : {}),
    });

    const { handoff, privateKey, publicKeyMultibase } = bootstrap;

    const resolver: Resolver = createResolver();
    const keyStore = createInMemoryKeyStore(handoff.agent_did, privateKey);
    const cedarSchemaJson = loadCedarSchema();
    const origin = originOf(handoff.well_known_urls.arp);

    const inboundHandlerHolder: { current: InboundHandler | null } = {
      current: options.onIncoming ?? null,
    };

    const dispatch: DispatchHandler = async (input: DispatchInput) => {
      const handler = inboundHandlerHolder.current;
      if (!handler) return {};
      const body = (input.message.body ?? {}) as Record<string, unknown>;
      const action = typeof body['action'] === 'string'
        ? body['action']
        : deriveActionFromType(input.message.type);
      const resource = coerceResource(body['resource']);
      const context = isPlainObject(body['context'])
        ? (body['context'] as Record<string, unknown>)
        : {};

      const reply = await handler(
        {
          action,
          resource,
          context,
          body,
          messageId: input.message.id,
          threadId: input.message.thid ?? null,
        },
        {
          connectionId: input.connectionId,
          peerDid: input.meta.peerDid,
          decision: {
            decision: input.decision.decision,
            obligations: input.decision.obligations,
            policiesFired: input.decision.policies_fired,
          },
          memory: input.memory,
          connection: input.connection,
          message: input.message,
          meta: input.meta,
        },
      );

      if (!reply || !reply.body) return {};

      // Auto-apply obligations to the outbound reply — developers never need
      // to call `egress()` on the inbound-reply path.
      const filtered = applyObligations(reply.body, input.decision.obligations);
      const out: { reply?: Record<string, unknown>; replyType?: string } = {};
      if (isPlainObject(filtered)) {
        out.reply = filtered as Record<string, unknown>;
      } else {
        out.reply = { value: filtered };
      }
      if (reply.replyType) out.replyType = reply.replyType;
      return out;
    };

    const runtime = await createRuntime({
      config: {
        did: handoff.agent_did,
        principalDid: handoff.principal_did,
        publicKeyMultibase,
        agentName: options.agentName ?? deriveAgentName(handoff.agent_did),
        agentDescription: options.agentDescription ?? 'ARP agent',
        wellKnownUrls: {
          didcomm: `${origin}/didcomm`,
          agentCard: handoff.well_known_urls.agent_card,
          arpJson: handoff.well_known_urls.arp,
        },
        representationVcUrl: `${origin}/.well-known/representation.jwt`,
        scopeCatalogVersion: options.scopeCatalogVersion ?? 'v1',
        tlsFingerprint: 'sdk-local-placeholder',
      },
      keyStore,
      resolver,
      ...(options.transportResolver
        ? { transportResolver: options.transportResolver }
        : {}),
      ...(options.transportFetch ? { transportFetch: options.transportFetch } : {}),
      cedarSchemaJson,
      registryPath: join(dataDir, 'registry.sqlite'),
      auditDir: join(dataDir, 'audit'),
      mailboxPath: join(dataDir, 'mailbox.sqlite'),
      ...(options.adminToken ? { adminToken: options.adminToken } : {}),
      dispatch,
      now,
    });

    const agent = new ArpAgent({
      did: handoff.agent_did,
      principalDid: handoff.principal_did,
      publicKeyMultibase,
      runtime,
      now,
    });
    // Agent exposes `onIncoming` to allow late-binding; rebind through holder.
    (agent as ArpAgent & { __inboundHandlerHolder: typeof inboundHandlerHolder }).__inboundHandlerHolder = inboundHandlerHolder;
    return agent;
  }

  /** Start the HTTP server. Returns the bound address. */
  async start(opts: { port?: number; host?: string } = {}): Promise<{ port: number; hostname: string }> {
    const port = opts.port ?? 4500;
    const host = opts.host ?? '127.0.0.1';
    return this.runtime.start(port, host);
  }

  /** Graceful shutdown. */
  async stop(opts: { graceMs?: number } = {}): Promise<void> {
    await this.runtime.stop(opts);
  }

  /**
   * Register the inbound handler late. When the agent was built without an
   * `onIncoming` option, adapters can attach one before calling `start()`.
   */
  onIncoming(handler: InboundHandler): void {
    const holder = (this as unknown as {
      __inboundHandlerHolder: { current: InboundHandler | null };
    }).__inboundHandlerHolder;
    if (!holder) {
      throw new Error('inbound handler holder missing — bug in ArpAgent.fromHandoff');
    }
    holder.current = handler;
  }

  /* --------------------- integration-point 1: check --------------------- */

  async check(input: CheckInput): Promise<PdpDecision> {
    const record = await this.runtime.registry.getConnection(input.connectionId);
    if (!record) {
      return {
        decision: 'deny',
        obligations: [],
        policies_fired: [],
        reasons: [`unknown_connection:${input.connectionId}`],
      };
    }
    if (record.status !== 'active') {
      return {
        decision: 'deny',
        obligations: [],
        policies_fired: [],
        reasons: [`connection_${record.status}`],
      };
    }
    if (await this.runtime.registry.isRevoked('connection', input.connectionId)) {
      return {
        decision: 'deny',
        obligations: [],
        policies_fired: [],
        reasons: ['revoked'],
      };
    }

    const principal = input.principal ?? {
      type: 'Agent',
      id: this.did,
      attrs: {
        connection_id: input.connectionId,
        owner_did: record.token.issuer,
      },
    };

    const decision = this.runtime.pdp.evaluate({
      cedarPolicies: record.cedar_policies,
      principal,
      action: input.action,
      resource: input.resource,
      context: input.context ?? {},
    });

    // Same pattern as the runtime: merge the token-level static obligations
    // with anything the PDP returned.
    const merged: Obligation[] = [
      ...(record.token.obligations ?? []),
      ...decision.obligations,
    ];
    return { ...decision, obligations: merged };
  }

  /* --------------------- integration-point 2: egress -------------------- */

  async egress(input: EgressInput): Promise<unknown> {
    let obligations = input.obligations;
    if (!obligations) {
      const record = await this.runtime.registry.getConnection(input.connectionId);
      obligations = record?.token.obligations ?? [];
    }
    return applyObligations(input.data, obligations);
  }

  /* --------------------- integration-point 3: audit --------------------- */

  async audit(event: AuditEventInput): Promise<void> {
    const log = this.runtime.auditFor(event.connectionId);
    // The audit package's schema enforces decision ∈ {allow, deny}. For
    // arbitrary developer event markers we downgrade to `allow` and encode
    // the intent in `reason` (keeps the chain valid without fragmenting the
    // schema across packages).
    const decision: 'allow' | 'deny' =
      event.decision === 'deny' ? 'deny' : 'allow';
    const reasonParts: string[] = [];
    if (event.decision && event.decision !== 'allow' && event.decision !== 'deny') {
      reasonParts.push(`event:${event.decision}`);
    }
    if (event.reason) reasonParts.push(event.reason);
    if (event.metadata) {
      reasonParts.push(`meta=${JSON.stringify(event.metadata)}`);
    }
    log.append({
      msg_id: event.messageId ?? `sdk_${this.now()}`,
      decision,
      policies_fired: event.policiesFired ?? [],
      ...(event.obligations && event.obligations.length > 0
        ? { obligations: event.obligations }
        : {}),
      ...(reasonParts.length > 0 ? { reason: reasonParts.join('; ') } : {}),
    });
  }

  /* ------------------- integration-point 4: lifecycle ------------------- */

  on<E extends ArpAgentEvent>(
    event: E,
    handler: (payload: ArpAgentEventPayload<E>) => void,
  ): void {
    // Typescript narrows poorly across the mapped type here; cast to the
    // concrete array and push. Runtime type is sound because caller must
    // name the event literal.
    (this.events[event] as Array<(e: ArpAgentEventPayload<E>) => void>).push(handler);
  }

  /** Emit a rotation event. Adapters call this when they detect key rotation. */
  emitRotation(event: RotationEvent): void {
    for (const h of this.events.rotation) h(event);
  }

  /* ------------------------- introspection ----------------------------- */

  /** Well-known docs the runtime publishes. */
  get wellKnown() {
    return this.runtime.wellKnown;
  }

  /** True when `stop()` has been called. */
  isDraining(): boolean {
    return this.runtime.isDraining();
  }

  inFlightCount(): number {
    return this.runtime.inFlightCount();
  }
}

/* --------------------------- API builders ---------------------------- */

function buildConnectionAPI(
  registry: Registry,
  runtime: Runtime,
  events: EventHandlers,
  now: () => number,
): ConnectionAPI {
  return {
    async list() {
      return registry.listConnections({ includeExpired: true });
    },
    async get(id) {
      return registry.getConnection(id);
    },
    async revoke(id, reason = 'sdk_revoked') {
      await runtime.revokeConnection(id, reason);
      for (const h of events.revocation) {
        h({ connectionId: id, reason, at: now() });
      }
    },
    async suspend(id) {
      await registry.updateStatus(id, 'suspended');
    },
    async resume(id) {
      await registry.updateStatus(id, 'active');
    },
    async add(token: ConnectionToken, tokenJws?: string): Promise<ConnectionRecord> {
      const record = await runtime.addConnection(token, tokenJws);
      for (const h of events.pairing) {
        h({
          connectionId: record.connection_id,
          peerDid: record.peer_did,
          at: now(),
        });
      }
      return record;
    },
  };
}

function buildRegistryReadAPI(registry: Registry): RegistryReadAPI {
  return {
    async get(id) {
      return registry.getConnection(id);
    },
    async list(opts) {
      return registry.listConnections(opts);
    },
  };
}

function buildPdpAPI(runtime: Runtime): PdpAPI {
  return {
    async check(input: CheckInput) {
      const record = await runtime.registry.getConnection(input.connectionId);
      if (!record) {
        return {
          decision: 'deny',
          obligations: [],
          policies_fired: [],
          reasons: [`unknown_connection:${input.connectionId}`],
        };
      }
      const principal = input.principal ?? {
        type: 'Agent',
        id: record.self_did,
        attrs: {
          connection_id: input.connectionId,
          owner_did: record.token.issuer,
        },
      };
      return runtime.pdp.evaluate({
        cedarPolicies: record.cedar_policies,
        principal,
        action: input.action,
        resource: input.resource,
        context: input.context ?? {},
      });
    },
  };
}

/* --------------------------- helpers --------------------------- */

function loadCedarSchema(): string {
  // We inline the read here so adapter authors don't need to know about
  // the schema path. Cedar schema file is shipped inside @kybernesis/arp-spec.
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const fs = requireFromHere('node:fs') as typeof import('node:fs');
  const path = requireFromHere.resolve('@kybernesis/arp-spec/cedar-schema.json');
  return fs.readFileSync(path, 'utf8');
}

function originOf(url: string): string {
  const u = new URL(url);
  return `${u.protocol}//${u.host}`;
}

function deriveAgentName(did: string): string {
  const host = did.split(':')[2] ?? 'agent';
  const first = host.split('.')[0] ?? 'agent';
  return first.charAt(0).toUpperCase() + first.slice(1);
}

function deriveActionFromType(type: string): string {
  const idx = type.lastIndexOf('/');
  return idx >= 0 ? type.slice(idx + 1) : type;
}

function coerceResource(spec: unknown): Resource {
  if (typeof spec === 'string') {
    const [type = 'Resource', id = spec] = spec.split(':', 2);
    return { type, id };
  }
  if (spec && typeof spec === 'object') {
    const s = spec as Record<string, unknown>;
    if (typeof s['type'] === 'string' && typeof s['id'] === 'string') {
      return {
        type: s['type'] as string,
        id: s['id'] as string,
        ...(isPlainObject(s['attrs']) ? { attrs: s['attrs'] as Record<string, unknown> } : {}),
      };
    }
  }
  return { type: 'Resource', id: 'default' };
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: python/arp-sdk/arp_sdk/agent.py
"""
Python `ArpAgent` — public API parity with the TypeScript SDK.

**v0.1.0 scope.** The Python SDK is published as a scaffold with:

- A full public API surface that matches `@kybernesis/arp-sdk` one-for-one
  (so adapter authors can type-check against it).
- A pure-Python `check` / `egress` / `audit` pipeline using the same
  obligation engine as the TS SDK.
- **Stubs** for handoff bootstrap, DIDComm transport, and Cedar evaluation.

The Cedar engine + DIDComm transport will be wired in the Phase-6 follow-up
(v1.1). Until then, agents can use the Python SDK for in-process
ARP-guarded tool calls against a pre-seeded connection registry (tests,
CI harnesses, local development). For full DIDComm interoperability use
the TypeScript SDK, or front the Python agent with the sidecar
(`@kybernesis/arp-sidecar`) via Mode B (see
`docs/ARP-installation-and-hosting.md §3.2`).
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Awaitable, Callable, Literal, Optional

from .obligations import apply_obligations
from .types import (
    AuditEvent,
    CheckInput,
    EgressInput,
    HandoffBundle,
    InboundHandler,
    Obligation,
    PairingEvent,
    PdpDecision,
    Resource,
    RevocationEvent,
    RotationEvent,
)


@dataclass
class _ConnectionRecord:
    connection_id: str
    peer_did: str
    status: Literal["active", "suspended", "revoked"] = "active"
    obligations: list[Obligation] = field(default_factory=list)
    cedar_policies: list[str] = field(default_factory=list)


class ArpAgent:
    """Developer-facing ARP agent (Python).

    Mirrors the public surface of the TypeScript `ArpAgent`. See
    `docs/ARP-installation-and-hosting.md §8` for the five integration
    points (`check`, `egress`, `on_incoming`, `audit`, `on`).
    """

    def __init__(
        self,
        *,
        did: str,
        principal_did: str,
        public_key_multibase: str,
        data_dir: Optional[str] = None,
    ) -> None:
        self.did = did
        self.principal_did = principal_did
        self.public_key_multibase = public_key_multibase
        self.data_dir = Path(data_dir) if data_dir else Path(".arp-data")
        self._connections: dict[str, _ConnectionRecord] = {}
        self._audit: dict[str, list[dict[str, Any]]] = {}
        self._inbound_handler: Optional[InboundHandler] = None
        self._revocation_handlers: list[Callable[[RevocationEvent], None]] = []
        self._rotation_handlers: list[Callable[[RotationEvent], None]] = []
        self._pairing_handlers: list[Callable[[PairingEvent], None]] = []

    # ------------------------------------------------------------------
    # Factory
    # ------------------------------------------------------------------
    @classmethod
    async def from_handoff(
        cls,
        handoff: "HandoffBundle | str | dict[str, Any]",
        *,
        on_incoming: Optional[InboundHandler] = None,
        data_dir: Optional[str] = None,
    ) -> "ArpAgent":
        bundle = _load_handoff(handoff)
        agent = cls(
            did=bundle.agent_did,
            principal_did=bundle.principal_did,
            public_key_multibase=bundle.public_key_multibase,
            data_dir=data_dir,
        )
        if on_incoming is not None:
            agent._inbound_handler = on_incoming
        return agent

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------
    async def start(self, *, port: int = 4500, host: str = "127.0.0.1") -> dict[str, Any]:
        """Start the HTTP DIDComm listener.

        v0.1.0: no-op (the Python SDK does not yet bind a DIDComm server;
        see module docstring). Returns placeholder metadata so the API
        shape matches the TS SDK.
        """
        return {"hostname": host, "port": port, "note": "arp_sdk v0.1.0 scaffold — HTTP listener not yet bound"}

    async def stop(self, *, grace_ms: int = 5000) -> None:
        return None

    def on_incoming(self, handler: InboundHandler) -> None:
        self._inbound_handler = handler

    # ------------------------------------------------------------------
    # 5 integration points
    # ------------------------------------------------------------------
    async def check(
        self,
        *,
        action: str,
        resource: Resource | dict[str, Any],
        connection_id: str,
        context: Optional[dict[str, Any]] = None,
    ) -> PdpDecision:
        _ = CheckInput(
            action=action,
            resource=_coerce_resource(resource),
            connection_id=connection_id,
            context=context or {},
        )
        record = self._connections.get(connection_id)
        if record is None:
            return PdpDecision(
                decision="deny",
                reasons=[f"unknown_connection:{connection_id}"],
            )
        if record.status != "active":
            return PdpDecision(
                decision="deny",
                reasons=[f"connection_{record.status}"],
            )
        # v0.1.0: allow-all, pass static token obligations through. Cedar
        # binding lands in v1.1 — at that point this switches to a real
        # policy evaluation. API surface stays identical.
        return PdpDecision(
            decision="allow",
            obligations=list(record.obligations),
            policies_fired=[p for p in record.cedar_policies if p],
        )

    async def egress(
        self,
        *,
        data: Any,
        connection_id: str,
        obligations: Optional[list[Obligation]] = None,
    ) -> Any:
        _ = EgressInput(data=data, connection_id=connection_id, obligations=obligations)
        if obligations is None:
            record = self._connections.get(connection_id)
            obligations = list(record.obligations) if record else []
        return apply_obligations(data, obligations)

    async def audit(
        self,
        *,
        connection_id: str,
        decision: str = "event",
        reason: Optional[str] = None,
        policies_fired: Optional[list[str]] = None,
        obligations: Optional[list[Obligation]] = None,
        metadata: Optional[dict[str, Any]] = None,
        message_id: Optional[str] = None,
    ) -> None:
        entry = {
            "msg_id": message_id or f"py_{len(self._audit.get(connection_id, []))}",
            "decision": decision,
            "reason": reason,
            "policies_fired": policies_fired or [],
            "obligations": [{"type": o.type, "params": o.params} for o in (obligations or [])],
            "metadata": metadata or {},
        }
        self._audit.setdefault(connection_id, []).append(entry)

    def on(
        self,
        event: Literal["revocation", "rotation", "pairing"],
        handler: Callable[..., None],
    ) -> None:
        if event == "revocation":
            self._revocation_handlers.append(handler)
        elif event == "rotation":
            self._rotation_handlers.append(handler)
        elif event == "pairing":
            self._pairing_handlers.append(handler)

    # ------------------------------------------------------------------
    # Admin / testing helpers (wired up to mirror TS SDK's ConnectionAPI)
    # ------------------------------------------------------------------
    def seed_connection(
        self,
        connection_id: str,
        *,
        peer_did: str,
        obligations: Optional[list[Obligation]] = None,
        cedar_policies: Optional[list[str]] = None,
    ) -> None:
        self._connections[connection_id] = _ConnectionRecord(
            connection_id=connection_id,
            peer_did=peer_did,
            obligations=list(obligations or []),
            cedar_policies=list(cedar_policies or []),
        )
        import time

        at = time.time()
        for h in self._pairing_handlers:
            h(PairingEvent(connection_id=connection_id, peer_did=peer_did, at=at))

    def revoke_connection(self, connection_id: str, *, reason: str = "owner_revoked") -> None:
        rec = self._connections.get(connection_id)
        if rec:
            rec.status = "revoked"
        import time

        at = time.time()
        for h in self._revocation_handlers:
            h(RevocationEvent(connection_id=connection_id, reason=reason, at=at))

    def audit_log(self, connection_id: str) -> list[dict[str, Any]]:
        """Return the in-memory audit entries for the given connection."""
        return list(self._audit.get(connection_id, []))


# ----------------------------------------------------------------------
# Helpers
# ----------------------------------------------------------------------
def _load_handoff(input: Any) -> HandoffBundle:
    if isinstance(input, HandoffBundle):
        return input
    if isinstance(input, str):
        raw = json.loads(Path(input).read_text())
    elif isinstance(input, dict):
        raw = input
    else:
        raise TypeError(f"handoff must be path, dict, or HandoffBundle; got {type(input)!r}")
    _reject_forbidden_fields(raw)
    return HandoffBundle(
        agent_did=raw["agent_did"],
        principal_did=raw["principal_did"],
        public_key_multibase=raw["public_key_multibase"],
        well_known_urls=raw["well_known_urls"],
        dns_records_published=list(raw.get("dns_records_published", [])),
        cert_expires_at=raw["cert_expires_at"],
        bootstrap_token=raw.get("bootstrap_token", ""),
    )


def _reject_forbidden_fields(value: Any, path: str = "$") -> None:
    if not isinstance(value, dict):
        return
    for k, v in value.items():
        if k.startswith("private") or k.startswith("secret") or "private_key" in k.lower():
            raise ValueError(
                f"handoff bundle contains forbidden field \"{path}.{k}\"; "
                "private material must not ship in a handoff"
            )
        _reject_forbidden_fields(v, f"{path}.{k}")


def _coerce_resource(spec: Any) -> Resource:
    if isinstance(spec, Resource):
        return spec
    if isinstance(spec, str):
        parts = spec.split(":", 1)
        return Resource(type=parts[0] or "Resource", id=parts[-1] or spec)
    if isinstance(spec, dict):
        return Resource(
            type=str(spec.get("type", "Resource")),
            id=str(spec.get("id", "default")),
            attrs=spec.get("attrs"),
            parents=spec.get("parents"),
        )
    return Resource(type="Resource", id="default")


__all__ = ["ArpAgent"]

# Silence unused-import warnings for optional types above.
_ = (AuditEvent, Awaitable)
